draft-bbs-signatures.md

```diff
@@ -439,205 +452,269 @@ Procedure:
+10. generators =  (H_s || H_d || H_1 || ... || H_L)
 
-11. B = P1 + H_s * s + H_d * domain + H_1 * msg_1 + ... + H_L * msg_L
+11. domain = hash_to_scalar((PK || L || generators || Ciphersuite_ID || header), 1)
 
-12. r3 = r1 ^ -1 mod r
+12. (r1, r2, e~, r2~, r3~, s~) = hash_to_scalar(PRF(8*ceil(log2(r))), 6)
 
-13. A' = A * r1
+13. (m~_j1, ..., m~_jU) =  hash_to_scalar(PRF(8*ceil(log2(r))), U)
 
-14. Abar = A' * (-e) + B * r1
+14. B = P1 + H_s * s + H_d * domain + H_1 * msg_1 + ... + H_L * msg_L
 
-15. D = B * r1 + H_s * r2
+15. r3 = r1 ^ -1 mod r
 
-16. s' = s + r2 * r3
+16. A' = A * r1
 
-17. C1 = A' * e~ + H_s * r2~
+17. Abar = A' * (-e) + B * r1
 
-18. C2 = D * (-r3~) + H_s * s~ + H_j1 * m~_j1 + ... + H_jU * m~_jU
+18. D = B * r1 + H_s * r2
 
-19. c = hash_to_scalar((PK || Abar || A' || D || C1 || C2 || ph), 1)
+19. s' = s + r2 * r3
 
-20. e^ = e~ + c * e
+20. C1 = A' * e~ + H_s * r2~
 
-21. r2^ = r2~ + c * r2
+21. C2 = D * (-r3~) + H_s * s~ + H_j1 * m~_j1 + ... + H_jU * m~_jU
 
-22. r3^ = r3~ + c * r3
+22. c = hash_to_scalar((PK || Abar || A' || D || C1 || C2 || ph), 1)
 
-23. s^ = s~ + c * s'
+23. e^ = e~ + c * e
 
-24. for j in (j1, j2,..., jU): m^_j = m~_j + c * msg_j
+24. r2^ = r2~ + c * r2
 
-25. proof = (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU))
+25. r3^ = r3~ + c * r3
 
-26. return proof
+26. s^ = s~ + c * s'
+
+27. for j in (j1, j2,..., jU): m^_j = m~_j + c * msg_j
+
+28. proof = (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU))
+
+29. return proof
 ```
 
 ### ProofVerify
 
-This operation checks that a proof is valid for a header, vector of revealed messages (along side their index corresponding to their original position when signed) and presentation header against a public key (PK).
+This operation checks that a proof is valid for a header, vector of generators, revealed messages, and presentation header against a public key (PK).
+
+The generators are supplied as a list (called `generators`) of points in G1. The generators MUST be the same as the ones supplied to the [Sign](#sign) operation, as part of the (generator, message) pairs of the `to_sign` input list. The generators list MUST also be sorted based on the order that the (generator, message) pairs had in that list. More specifically, if the list `to_sign = [(H_1, msg_1), (H_2, msg_2), ..., (H_L, msg_L)]` was supplied to the Sign operation as input, then the generators list supplied as input to ProoVerify MUST equal `[H_1, H_2, ..., H_L]`, in that order.
+
+The revealed messages are supplied as a list of at most `length(generators)` pairs, of the form (revealed_message, index), containing a non-zero scalar mod r, representing a (revealed) message, and an integer in the 1 to L range, representing the index that the (generator, message) pair containing the revealed_message had in the `to_sign` list. The list MUST be sorted based on the (ascending) order of the indexes. If the list is not supplied, it will default to the empty array ("[]").
```

Comment:
It seems a bit more natural to use (index, message) to me given that in many languages that makes the list naturally sortable, sorting by the first and then the second element